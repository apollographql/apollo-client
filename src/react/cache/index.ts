export type { SuspenseCacheOptions } from './SuspenseCache.js';
export { getSuspenseCache } from './getSuspenseCache.js';

import { SuspenseCache as RealSuspenseCache } from './SuspenseCache.js';

// TODO: remove export with release 3.8
// replace with
// export type { SuspenseCache } from './SuspenseCache.js';
/**
 * @deprecated 
 * It is no longer necessary to create a `SuspenseCache` instance and pass it into the `ApolloProvider`.
 * Please remove this code from your application.
 *  
 * This export will be removed with the final 3.8 release.
 */
export class SuspenseCache extends RealSuspenseCache {
  constructor(){
    super();
    // throwing an error here instead of using invariant - we do not want this error
    // message to be link-ified, but to directly show up as bold as possible
    throw new Error(
      'It is no longer necessary to create a `SuspenseCache` instance and pass it into the `ApolloProvider`.\n' + 
      'Please remove this code from your application. \n\n' + 
      'This export will be removed with the final 3.8 release.'
    );
  }
}