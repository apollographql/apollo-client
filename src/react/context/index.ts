export * from './ApolloConsumer';
export * from './ApolloContext';
export * from './ApolloProvider';
