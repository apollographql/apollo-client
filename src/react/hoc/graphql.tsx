import { DocumentNode } from 'graphql';

import { parser, DocumentType } from '../parser';
import { withQuery } from './query-hoc';
import { withMutation } from './mutation-hoc';
import { withSubscription } from './subscription-hoc';
import { OperationOption, DataProps, MutateProps } from './types';

export function graphql<
  TProps extends TGraphQLVariables | {} = {},
  TData = {},
  TGraphQLVariables = {},
  TChildProps = Partial<DataProps<TData, TGraphQLVariables>> &
    Partial<MutateProps<TData, TGraphQLVariables>>
>(
  document: DocumentNode,
  operationOptions: OperationOption<
    TProps,
    TData,
    TGraphQLVariables,
    TChildProps
  > = {}
): (
  WrappedComponent: React.ComponentType<TProps & TChildProps>
) => React.ComponentClass<TProps> {
  switch (parser(document).type) {
    case DocumentType.Mutation:
      return withMutation(document, operationOptions);
    case DocumentType.Subscription:
      return withSubscription(document, operationOptions);
    case DocumentType.Query:
    default:
      return withQuery(document, operationOptions);
  }
}
