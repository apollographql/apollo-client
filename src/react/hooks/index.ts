export * from './useApolloClient';
export * from './useLazyQuery';
export * from './useMutation';
export * from './useQuery';
export * from './useSubscription';
