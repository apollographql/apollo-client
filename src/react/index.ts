import { invariant } from "ts-invariant";
import { DEV } from "../utilities";
invariant("boolean" === typeof DEV, DEV);

export {
  ApolloProvider,
  ApolloConsumer,
  getApolloContext,
  resetApolloContext,
  ApolloContextValue
} from './context';

export * from './hooks';

export {
  DocumentType,
  IDocumentDefinition,
  operationName,
  parser
} from './parser';

export * from './types/types';
