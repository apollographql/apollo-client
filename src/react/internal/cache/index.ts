export type { SuspenseCacheOptions } from "./SuspenseCache.js";
export type { CacheKey, QueryKey } from "./types.js";
export { getSuspenseCache } from "./getSuspenseCache.js";
export type { QueryReference, QueryRefPromise } from "./QueryReference.js";
export {
  InternalQueryReference,
  unwrapQueryRef,
  wrapQueryRef,
  updateWrappedQueryRef,
  getWrappedPromise,
} from "./QueryReference.js";
