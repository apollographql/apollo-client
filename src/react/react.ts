let React: typeof import('react');

// Apollo Client can be used without React, which means we want to make sure
// `react` is only imported/required if actually needed. To help with this
// the `react` module is lazy loaded using `requireReactLazily` when used by
// Apollo Client's React integration layer.
export function requireReactLazily(): typeof import('react') {
  return React || (React = require('react'));
}
