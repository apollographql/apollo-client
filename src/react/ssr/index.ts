export { getMarkupFromTree, getDataFromTree } from './getDataFromTree';
export { renderToStringWithData } from './renderToStringWithData';
export { RenderPromises } from './RenderPromises';
