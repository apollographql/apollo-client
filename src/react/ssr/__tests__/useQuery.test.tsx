import React from 'react';
import { DocumentNode } from 'graphql';
import gql from 'graphql-tag';
import { MockedProvider, mockSingleLink } from '../../../testing';
import { ApolloClient } from '../../../core';
import { InMemoryCache } from '../../../cache';
import { ApolloProvider } from '../../context';
import { useApolloClient, useQuery } from '../../hooks';
import { render, wait } from '@testing-library/react';
import { renderToStringWithData } from '..';

describe('useQuery Hook SSR', () => {
  const CAR_QUERY: DocumentNode = gql`
    query {
      cars {
        make
        model
        vin
      }
    }
  `;

  const CAR_RESULT_DATA = {
    cars: [
      {
        make: 'Audi',
        model: 'RS8',
        vin: 'DOLLADOLLABILL',
        __typename: 'Car'
      }
    ]
  };

  const CAR_MOCKS = [
    {
      request: {
        query: CAR_QUERY
      },
      result: { data: CAR_RESULT_DATA }
    }
  ];

  it('should support SSR', () => {
    const Component = () => {
      const { loading, data } = useQuery(CAR_QUERY);
      if (!loading) {
        expect(data).toEqual(CAR_RESULT_DATA);
        const { make, model, vin } = data.cars[0];
        return (
          <div>
            {make}, {model}, {vin}
          </div>
        );
      }
      return null;
    };

    const app = (
      <MockedProvider mocks={CAR_MOCKS}>
        <Component />
      </MockedProvider>
    );

    return renderToStringWithData(app).then(markup => {
      expect(markup).toMatch(/Audi/);
    });
  });

  it('should initialize data as `undefined` when loading', () => {
    const Component = () => {
      const { data, loading } = useQuery(CAR_QUERY);
      if (loading) {
        expect(data).toBeUndefined();
      }
      return null;
    };

    const app = (
      <MockedProvider mocks={CAR_MOCKS}>
        <Component />
      </MockedProvider>
    );

    return renderToStringWithData(app);
  });

  it('should skip SSR tree rendering if `ssr` option is `false`', async () => {
    let renderCount = 0;
    const Component = () => {
      const { data, loading } = useQuery(CAR_QUERY, { ssr: false });
      renderCount += 1;

      if (!loading) {
        const { make } = data.cars[0];
        return <div>{make}</div>;
      }
      return null;
    };

    const app = (
      <MockedProvider mocks={CAR_MOCKS}>
        <Component />
      </MockedProvider>
    );

    return renderToStringWithData(app).then(result => {
      expect(renderCount).toBe(1);
      expect(result).toEqual('');
    });
  });

  it(
    'should skip both SSR tree rendering and SSR component rendering if ' +
      '`ssr` option is `false` and `ssrMode` is `true`',
    async () => {
      const link = mockSingleLink({
        request: { query: CAR_QUERY },
        result: { data: CAR_RESULT_DATA }
      });

      const client = new ApolloClient({
        cache: new InMemoryCache(),
        link,
        ssrMode: true
      });

      let renderCount = 0;
      const Component = () => {
        const { data, loading } = useQuery(CAR_QUERY, { ssr: false });

        let content = null;
        switch (renderCount) {
          case 0:
            expect(loading).toBeTruthy();
            expect(data).toBeUndefined();
            break;
          case 1: // FAIL; should not render a second time
          default:
        }

        renderCount += 1;
        return content;
      };

      const app = (
        <ApolloProvider client={client}>
          <Component />
        </ApolloProvider>
      );

      await renderToStringWithData(app).then(result => {
        expect(renderCount).toBe(1);
        expect(result).toEqual('');
      });

      renderCount = 0;

      render(
        <ApolloProvider client={client}>
          <Component />
        </ApolloProvider>
      );

      await wait(() => {
        expect(renderCount).toBe(1);
      });
    }
  );

  it('should skip SSR tree rendering if `skip` option is `true`', async () => {
    let renderCount = 0;
    const Component = () => {
      const {
        loading,
        networkStatus,
        data,
      } = useQuery(CAR_QUERY, { skip: true });
      renderCount += 1;

      expect(loading).toBeFalsy();
      expect(networkStatus).toBe(7);
      expect(data).toBeUndefined();

      return null;
    };

    const app = (
      <MockedProvider mocks={CAR_MOCKS}>
        <Component />
      </MockedProvider>
    );

    return renderToStringWithData(app).then(result => {
      expect(renderCount).toBe(1);
      expect(result).toBe('');
    });
  });

  it('should return data written previously to cache during SSR pass if using cache-only fetchPolicy', async () => {
    const cache = new InMemoryCache({
      typePolicies: {
        Order: {
          keyFields: ["selection"],
        },
      },
    });

    const query = gql`
      query GetSearchResults {
        getSearchResults @client {
          locale
          order {
            selection
          }
          pagination {
            pageLimit
          }
          results {
            id
            text
          }
        }
      }
    `;

    const initialData = {
      getSearchResults: {
        __typename: 'SearchResults',
        locale: 'en-US',
        order: {
          __typename: 'Order',
          selection: 'RELEVANCE',
        },
        pagination: {
          pageLimit: 3,
        },
        results: [
          { __typename: "SearchResult", id: 1, text: "hi" },
          { __typename: "SearchResult", id: 2, text: "hello" },
          { __typename: "SearchResult", id: 3, text: "hey" },
        ],
      },
    };

    const spy = jest.fn();

    const Component = () => {
      useApolloClient().writeQuery({ query, data: initialData });;

      const { loading, data } = useQuery(query, {
        fetchPolicy: 'cache-only',
      });

      spy(loading);

      if (!loading) {
        expect(data).toEqual(initialData);

        const {
          getSearchResults: {
            pagination: {
              pageLimit,
            },
          },
        } = data;
        return (
          <div>
            {pageLimit}
          </div>
        );
      }
      return null;
    };

    const app = (
      <MockedProvider
        addTypename
        cache={cache}
      >
        <Component />
      </MockedProvider>
    );

    return renderToStringWithData(app).then(markup => {
      expect(spy).toHaveBeenNthCalledWith(1, false);
      expect(markup).toMatch(/<div.*>3<\/div>/);
      expect(cache.extract()).toMatchSnapshot();
    });
  });
});
