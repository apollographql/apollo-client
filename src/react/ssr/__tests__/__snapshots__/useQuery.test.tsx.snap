// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useQuery Hook SSR should return data written previously to cache during SSR pass if using cache-only fetchPolicy 1`] = `
Object {
  "Order:{\\"selection\\":\\"RELEVANCE\\"}": Object {
    "__typename": "Order",
    "selection": "RELEVANCE",
  },
  "ROOT_QUERY": Object {
    "__typename": "Query",
    "getSearchResults": Object {
      "__typename": "SearchResults",
      "locale": "en-US",
      "order": Object {
        "__ref": "Order:{\\"selection\\":\\"RELEVANCE\\"}",
      },
      "pagination": Object {
        "pageLimit": 3,
      },
      "results": Array [
        Object {
          "__ref": "SearchResult:1",
        },
        Object {
          "__ref": "SearchResult:2",
        },
        Object {
          "__ref": "SearchResult:3",
        },
      ],
    },
  },
  "SearchResult:1": Object {
    "__typename": "SearchResult",
    "id": 1,
    "text": "hi",
  },
  "SearchResult:2": Object {
    "__typename": "SearchResult",
    "id": 2,
    "text": "hello",
  },
  "SearchResult:3": Object {
    "__typename": "SearchResult",
    "id": 3,
    "text": "hey",
  },
}
`;
